### PR TITLE
Bump regalloc2 to 0.7.0

### PR DESCRIPTION
**Pull request by elliottt**
Bump regalloc2 to version 0.7.0, which removes program moves and adds back in support for scratch registers. This bump also brings with it a new dependency on `hashbrown-0.13.2` and `rustc-hash`. The latter had an audit that we could import, while the former was updated in #6238.

### CODE REVIEW

**Comment by cfallin** [comment]
This will need the `is_move` method removed from the `Function` trait impl on `VCode`, and I think also the corresponding if-statement to skip emitting program moves removed from the VCode emission path ([here](https://github.com/bytecodealliance/wasmtime/blob/a486aa37ad187e4b5c5d34da303a39fa5cdd2745/cranelift/codegen/src/machinst/vcode.rs#L1285) and [here]https://github.com/bytecodealliance/wasmtime/blob/a486aa37ad187e4b5c5d34da303a39fa5cdd2745/cranelift/codegen/src/machinst/vcode.rs#L929) respectively, and the `is_move` data structure DCE'd too if it's now dead).

**Comment by cfallin** [comment]
> I think it is unused now, nice! That allows us to remove insertions into the map during `collect_operands`, and the only use of `MachInst::is_move` is now for that conditional that guards the two assertions about vregs with moves. I'm inclined to leave that as-is, as I'd like to keep that assertion around, what do you think?

You mean [this](https://github.com/bytecodealliance/wasmtime/blob/51ed20ab4d855343ee68b46b5aa98002c5c63dca/cranelift/codegen/src/machinst/vcode.rs#L571-L593) block of code? That's fine to keep I think, yeah; the `is_move` at the top there is an invocation of `MachInst::is_move()`, which is separate from `<VCode as regalloc2::Function>::is_move()` and is just a match on the instruction. In other words we can remove the `is_move` hashmap while still keeping this check.

**Comment by elliottt** [comment]
> You mean [this](https://github.com/bytecodealliance/wasmtime/blob/51ed20ab4d855343ee68b46b5aa98002c5c63dca/cranelift/codegen/src/machinst/vcode.rs#L571-L593) block of code? That's fine to keep I think, yeah; the `is_move` at the top there is an invocation of `MachInst::is_move()`, which is separate from `<VCode as regalloc2::Function>::is_move()` and is just a match on the instruction. In other words we can remove the `is_move` hashmap while still keeping this check.

Yep, I agree with that. I think we could also remove the `MachInst::is_move` function if we wanted to lose those two assertions. The last two `let` decls in that block go away as dead code once we remove the table, which is what had me wondering if we still needed `MachInst::is_move` at all at this point.

File: cranelift/filetests/filetests/isa/x64/struct-ret.clif [hunk 1]
```diff
@@ -95,7 +96,8 @@ block0(v0: i64):
 ;   subq $0x10, %rsp
 ;   movq %r15, (%rsp)
 ; block1: ; offset 0xc
-;   movq %rdi, %r15
+;   movq %rdi, %rax
+;   movq %rax, %r15
```
Comment (elliottt) [inline cranelift/filetests/filetests/isa/x64/struct-ret.clif:100]: This is a little surprising, but also a totally reasonable consequence of no longer communicating program moves to RA2.

**Comment by jameysharp** [comment]
It looks to me like ABI register constraints on incoming args and outgoing return values are now emitting moves, more or less everywhere. Am I interpreting the filetest changes correctly, and is that an expected result? I thought those were all implemented as register constraints on `ret` and the `args` pseudo-instruction, not as actual move instructions, so I thought RA2 would still understand how to deal with them.

**Comment by elliottt** [comment]
> It looks to me like ABI register constraints on incoming args and outgoing return values are now emitting moves, more or less everywhere. Am I interpreting the filetest changes correctly, and is that an expected result? I thought those were all implemented as register constraints on `ret` and the `args` pseudo-instruction, not as actual move instructions, so I thought RA2 would still understand how to deal with them.

I think it's not so much that they're emitting moves, but that regalloc2 is no longer removing them. This is more or less expected, but I'd like to do a little investigation to see if we can improve things before we merge this.

**Comment by cfallin** [comment]
> I think it's not so much that they're emitting moves, but that regalloc2 is no longer removing them. This is more or less expected, but I'd like to do a little investigation to see if we can improve things before we merge this.

I would say this is very much *not* expected, and we should not merge with this change. It's pretty significant if we can no longer use the constraints to refer to the args in their original values.

I'd have expected everything to work, because the way that we handle args now is with an [args pseudo-inst](https://github.com/bytecodealliance/wasmtime/blob/ef7af28ef095277f5979dcbc6ff03964c1e0ba12/cranelift/codegen/src/machinst/abi.rs#L429-L431) that defs all the argument vregs with constraints as the very first instruction in the function.

**Comment by cfallin** [comment]
(And, as a reward for skimming quickly, I failed to notice that @jameysharp made exactly the same point two comments earlier -- sorry! Yes, we need to fix this before merging.)

**Comment by elliottt** [comment]
> I would say this is very much _not_ expected, and we should not merge with this change. It's pretty significant if we can no longer use the constraints to refer to the args in their original values.

I didn't mean to imply that we should merge with this behavior, just that this output made sense to me given that program moves had been removed in the new version of RA2.

**Comment by cfallin** [comment]
> > I would say this is very much _not_ expected, and we should not merge with this change. It's pretty significant if we can no longer use the constraints to refer to the args in their original values.
> 
> I didn't mean to imply that we should merge with this behavior, just that this output made sense to me given that program moves had been removed in the new version of RA2.

Right, but the current version of the ABI code doesn't use moves, so it's unexpected that anything should change I think.

**Comment by cfallin** [comment]
After a little offline discussion with @elliottt it seems what may be happening is that some remaining program-level moves *elsewhere* in Cranelift's output (i.e., not in ABI code) were still present, and no longer handling these at the regalloc level is what's causing the regressions now. These didn't necessarily go away during SSA-ification because a move is a perfectly reasonable SSA op (namely, it's the identity function). We'll want to remove the last few of these before moving to RA2 0.7.0; it shouldn't be a very large effort.

File: cranelift/filetests/filetests/isa/x64/struct-ret.clif (hunk 1 above)
Comment (elliottt) [reply cranelift/filetests/filetests/isa/x64/struct-ret.clif]: This was easily resolved by copying the sret register in `Lower::new`, instead of generating a temporary that we would move into later.

File: cranelift/filetests/filetests/isa/x64/struct-ret.clif [hunk 2]
```diff
@@ -96,8 +97,9 @@ block0(v0: i64):
 ;   movq %r15, (%rsp)
 ; block1: ; offset 0xc
 ;   movq %rdi, %r15
-;   movabsq $0, %rdx ; reloc_external Abs8 %f4 0
-;   callq *%rdx
+;   movabsq $0, %rax ; reloc_external Abs8 %f4 0
+;   movq %r15, %rdi
```
Comment (elliottt) [inline cranelift/filetests/filetests/isa/x64/struct-ret.clif:101]: This is a great example of something the redundant move eliminator would have fixed before.

File: cranelift/filetests/filetests/isa/s390x/bitops.clif [hunk 3]
```diff
@@ -353,11 +353,13 @@ block0(v0: i64):
 ; VCode:
 ; block0:
 ;   flogr %r2, %r2
+;   lgr %r2, %r2
 ;   br %r14
 ; 
 ; Disassembled:
 ; block0: ; offset 0x0
 ;   flogr %r2, %r2
+;   lgr %r2, %r2
```
Comment (alexcrichton) [inline cranelift/filetests/filetests/isa/s390x/bitops.clif:362]: Random drive-by comment, but this looks like something s390x-specific may be required here?

File: cranelift/filetests/filetests/isa/riscv64/call.clif [hunk 4]
```diff
@@ -396,11 +396,13 @@ block0(v0: i128, v1: i64):
...
-;   mv a0,a1
+;   mv a0,a2
+;   mv a1,a0
 ;   ret
 ; 
 ; Disassembled:
 ; block0: ; offset 0x0
+;   ori a2, a0, 0
 ;   ori a0, a1, 0
 ;   ret
```
Comment (jameysharp) [inline cranelift/filetests/filetests/isa/riscv64/call.clif:407]: Here the VCode pretty-printing changed without changing the disassembly listing. The same instruction was originally printed as `mv a0,a1` and now is printed as `mv a1,a0`, but in both cases it's disassembled as `ori a0, a1, 0`. Any idea what's going on there? I don't see any Rust-side changes that explain this. There are more cases like this below.

File: cranelift/filetests/filetests/isa/riscv64/reduce.clif [hunk 5]
```diff
@@ -10,10 +10,12 @@ block0(v0: i128):
 
 ; VCode:
 ; block0:
+;   mv a0,a0
```
Comment (jameysharp) [inline cranelift/filetests/filetests/isa/riscv64/reduce.clif:13]: I'm hoping that whatever is causing `mv a0,a0` in the riscv backend is easy to fix.

File: cranelift/filetests/filetests/isa/s390x/bitops.clif [hunk 6]
```diff
@@ -353,11 +353,13 @@ block0(v0: i64):
 ; VCode:
 ; block0:
 ;   flogr %r2, %r2
+;   lgr %r2, %r2
```
Comment (jameysharp) [inline cranelift/filetests/filetests/isa/s390x/bitops.clif:356]: Similarly, I hope whatever is causing `lgr %r2, %r2` in s390x is easy to fix.

File: cranelift/filetests/filetests/isa/aarch64/call.clif [hunk 7]
```diff
@@ -850,16 +850,14 @@ block0(v0: i64):
 
 ; VCode:
 ; block0:
-;   mov x5, x8
-;   movz x4, #42
-;   str x4, [x8]
+;   movz x2, #42
+;   str x2, [x8]
```
Comment (jameysharp) [inline cranelift/filetests/filetests/isa/aarch64/call.clif:854]: It's neat that this is actually removing moves in some cases compared to the older RA2!

Comment (elliottt) [reply cranelift/filetests/filetests/isa/aarch64/call.clif:854]: That's actually due to my change to `sret` register handling :sweat_smile: 

I'll land it as a separate PR once I've resolved all of the new move issues.

File: cranelift/filetests/filetests/isa/aarch64/reftypes.clif [hunk 8]
```diff
@@ -92,21 +92,23 @@ block3(v7: r64, v8: r64):
 ;   mov fp, sp
 ;   sub sp, sp, #32
 ; block0:
-;   str x0, [sp, #8]
+;   mov x3, x0
 ;   str x1, [sp, #16]
+;   mov x0, x3
+;   str x3, [sp, #8]
```
Comment (elliottt) [inline cranelift/filetests/filetests/isa/aarch64/reftypes.clif:98]: This pattern shows up across all backends. We insert a copy of the ref-typed values before calling a function with them as an argument, as the values themselves will be constrained to live on the stack.
https://github.com/bytecodealliance/wasmtime/blob/e6339b27254d9e9c4c2fe7bdf1a3a06589c74847/cranelift/codegen/src/machinst/abi.rs#L2243-L2253
With program moves no longer being special for RA2, the move that's inserted by cranelift on line 95 doesn't end up being redundant with the one that RA2 inserts. The store to the stack can use that value without any issue, however we still need to copy the value back to `x0` to pass as the first argument to `fn0`. Supporting overlap in RA2 will help a lot here, as the value will be allowed to exist on the stack and in a register at the same time, giving us a path forward for skipping the initial copy when setting up the arg.

**Review by alexcrichton — APPROVED** [review]

**Review by jameysharp — COMMENTED** [review]
Ooh, I wonder if this will work for me now:
/bench_x64

**Review by cfallin — APPROVED** [review]
Updates look good now, diffs are minimal and well-understood -- thanks for the diligence here!

**Comment by jlb6740** [comment]
compilation :: cycles :: benchmarks/bz2/benchmark.wasm

  Δ = 78635410.00 ± 70514488.64 (confidence = 99%)

  main.so is 1.03x to 1.47x faster than commit.so!

  [238904854 396078552.64 622424546] commit.so
  [218109284 317443142.64 529913904] main.so

instantiation :: cycles :: benchmarks/bz2/benchmark.wasm

  Δ = 49379.12 ± 40859.40 (confidence = 99%)

  main.so is 1.04x to 1.41x faster than commit.so!

  [178906 270460.48 449972] commit.so
  [167434 221081.36 300412] main.so

compilation :: cycles :: benchmarks/spidermonkey/benchmark.wasm

  No difference in performance.

  [7407499152 10121661523.12 17407002596] commit.so
  [7004620868 8377593595.52 9898764482] main.so

compilation :: cycles :: benchmarks/pulldown-cmark/benchmark.wasm

  No difference in performance.

  [308082960 444433126.32 1357616580] commit.so
  [300892950 372722175.60 532565542] main.so

execution :: cycles :: benchmarks/bz2/benchmark.wasm

  No difference in performance.

  [113777830 166523657.76 183949430] commit.so
  [109460994 144644223.28 188312718] main.so

instantiation :: cycles :: benchmarks/pulldown-cmark/benchmark.wasm

  No difference in performance.

  [201324 251931.76 408750] commit.so
  [187242 239093.04 361026] main.so

execution :: cycles :: benchmarks/pulldown-cmark/benchmark.wasm

  No difference in performance.

  [8336902 10949792.56 29211712] commit.so
  [8141614 10515012.32 15191794] main.so

execution :: cycles :: benchmarks/spidermonkey/benchmark.wasm

  No difference in performance.

  [1023392010 1488396907.12 2048405904] commit.so
  [994591294 1539428857.92 2059988336] main.so

instantiation :: cycles :: benchmarks/spidermonkey/benchmark.wasm

  No difference in performance.

  [494164 615861.28 799082] commit.so
  [489712 631441.20 863642] main.so

**Comment by elliottt** [comment]
The above numbers concerned me, so I ran the bz2 and spidermonkey benchmarks on dedicated hardware. They all showed `No difference`, so I'm going to attribute the wild difference above to running on CI infrastructure.
